aidial_assistant/tools_chain/tools_chain.py

```diff
@@ -0,0 +1,155 @@
+import json
+from typing import Any
+
+from aidial_sdk.chat_completion import Role
+from openai import BadRequestError
+
+from aidial_assistant.chain.callbacks.chain_callback import ChainCallback
+from aidial_assistant.chain.callbacks.command_callback import CommandCallback
+from aidial_assistant.chain.command_chain import CommandDict
+from aidial_assistant.chain.command_result import (
+    CommandInvocation,
+    CommandResult,
+    Status,
+    commands_to_text,
+    responses_to_text,
+)
+from aidial_assistant.chain.model_response_reader import (
+    AssistantProtocolException,
+)
+from aidial_assistant.commands.base import Command
+from aidial_assistant.model.model_client import (
+    ExtraResultsCallback,
+    Message,
+    ModelClient,
+    ToolCall,
+)
+from aidial_assistant.utils.open_ai import Tool
+
+
+def _publish_command(
+    command_callback: CommandCallback, name: str, arguments: str
+):
+    command_callback.on_command(name)
+    args_callback = command_callback.args_callback()
+    args_callback.on_args_start()
+    args_callback.on_args_chunk(arguments)
+    args_callback.on_args_end()
+
+
+class ToolCallsCallback(ExtraResultsCallback):
+    def __init__(self):
+        self.tool_calls: list[ToolCall] = []
+
+    def on_tool_calls(self, tool_calls: list[ToolCall]):
+        self.tool_calls = tool_calls
+
+
+class ToolsChain:
+    def __init__(
+        self,
+        model: ModelClient,
+        tools: list[Tool],
+        command_dict: CommandDict,
+    ):
+        self.model = model
+        self.tools = tools
+        self.command_dict = command_dict
```

Comment:
There is an implicit assumption here that `size(command_dict) == size(tools)` and moreover names of the tools must match names of the commands.

Why not merge them into a single list? `tools: dict[str, CommandTool]`, where `CommandTool = Tuple[Command, Tool]`

Reply:
Done.